### PR TITLE
Implement target backend for AIRBIN

### PR DESCRIPTION
**Pull request by jnider**
AIRBIN is a stand-alone file format that contains a series of configuration instructions for AIE. This new backend outputs an ELF-compatible file that contains all of the register values needed for configuration. The file is uploaded to the control processor that executes the series of commands, instead of the host executing each command one by one.

This depends on the 'elfutils' library.
https://github.com/jnider/elfutils.git

Configure elfutils with:
./configure --program-prefix="air-" --disable-debuginfod

Generate an AIRBIN by passing "--aie-generate-airbin" to aircc.py.

v2: rewrite macros using C++

### CODE REVIEW

**Comment by fifield** [comment]
I'm curious why we need to build libelf instead of `apt install libelf-dev` ?

**Comment by jnider** [comment]
> I'm curious why we need to build libelf instead of `apt install libelf-dev` ?

I have modified libelf to add a new type (EM_AMDAIR) so the generated ELF file can be uniquely identified as an AIRBIN: https://github.com/jnider/elfutils/commit/c0b50e79aef3fcb876b1828d456b09529dd9ad80

This change cannot yet be upstreamed to libelf because the future of AIRBIN is not settled. There are some ideas to change it to AIEBIN, or modify the contents to be a nested file containing CDOs (or similar).

**Comment by jnider** [comment]
Please approve so the workflows can run

**Comment by jnider** [comment]
Looks like I need approval again
It seems pkg-config is the culprit that was preventing elfutils from building. It is installed automatically on machines with ubuntu-desktop therefore it was hard to track down the difference from the CI installation.

**Comment by jnider** [comment]
Added ELFUTILS_DIR path to cmake in .github/workflows/buildAndTest.yml

**Comment by jnider** [comment]
Please approve for CI

I changed the $ELFUTILS variable (which was not defined anywhere) to `pwd`/elfutils, which should be the absolute path that contains the library.

**Review by stephenneuendorffer — APPROVED** [review]

**Comment by jnider** [comment]
Please approve the CI workflow

Changed `pwd`/elfutils into `pwd`/../elfutils - now I understand the .. means the build directory.

**Comment by stephenneuendorffer** [comment]
Depends on #570

**Comment by stephenneuendorffer** [comment]
> Having made that mistake and thus gotten a successful test going pretty smoothly because of being able to use the system package (by substituting `EM_AMDGPU`), I would **highly highly highly** recommend we do not use the fork, which implies taking on a source dependency (for ourselves and our users).

Since we already have an LLVM dependence, would it make sense to use the code there?  I'm not opposed to getting this checked in (as planned) and then figuring out how to do that.  I had somewhat assumed that the tooling would handle an unknown EM_ id just fine  (we did that in Peano for quite a while)

**Comment by makslevental** [comment]
@jnider for some reason your commit is lost in the void (because it's on a commit I overwrote with a `push -f`?) so I'm transcribing it here:

> Thanks for writing the test. I will consider your idea about just using a define for the ELF type. It does simplify the project but is really only a temporary hack. I guess it really depends on the future of the AIRBIN format. If we are serious about making it official, we will have to keep a fork of libelf until it can be properly merged upstream.

I'm curious why/what you mean by this. I looked at your fork and the delta and it seems to be only the `EM_AMDAIR <-> 255` mapping. I'm not familiar with ELF so I don't know what the e-machine id implies structurally about ELF itself. What I'm saying is can you tell me what libelf has to do differently to accomodate `EM_AMDAIR` than what you currently have in your fork (that would thus necessitate upstreaming)?

**Comment by jnider** [comment]
Sorry, I'm writing from my phone so the UI is not great...The general structure of ELF remains the same, but the number gets assigned by some official body (that I haven't quite figured out who's in charge yet) so it is internationally recognized and all implementations of ELF libraries will know its meaning. It then gives us the ability to define additional headers that are specific to this format. For example, if we decide to embed CDO files directly in the ELF, we could decide that is part of the AIRBIN format and have code in libelf to encode/decode them. If we just use a define, only our tools will recognize what it is and we run the risk of the same number being assigned to some other ELF type in the future, meaning official tools will view our AIRBIN files as some other kind of ELF architecture, at minimum identifying it incorrectly and potentially decoding any extensions incorrectly.So for now it's all the same (no one outside the company is aware of which number we have chosen) but that is likely to change as soon as people outside the company start to use AIRBIN.
-------- Original message --------From: Maksim Levental ***@***.***> Date: 2023-11-30  20:13  (GMT-08:00) To: Xilinx/mlir-aie ***@***.***> Cc: Joel Nider ***@***.***>, Mention ***@***.***> Subject: Re: [Xilinx/mlir-aie] Implement target backend for AIRBIN (PR #668) 
@jnider for some reason your commit is lost in the void (because it's on a commit I overwrote with a push -f?) so I'm transcribing it here:

Thanks for writing the test. I will consider your idea about just using a define for the ELF type. It does simplify the project but is really only a temporary hack. I guess it really depends on the future of the AIRBIN format. If we are serious about making it official, we will have to keep a fork of libelf until it can be properly merged upstream.

I'm curious why/what you mean by this. I looked at your fork and the delta and it seems to be only the EM_AMDAIR <-> 255 mapping. I'm not familiar with ELF so I don't know what the e-machine id implies structurally about ELF itself. What I'm saying is can you tell me what libelf has to do differently to accomodate EM_AMDAIR than what you currently have in your fork (that would thus necessitate upstreaming)?

**Comment by makslevental** [comment]
@jnider thanks for the explanation. I expected as much but wasn't sure. So are you comfortable with this approach until we get the as-of-yet-uknown standards body on board? If so, I'll keep going with my plans for the tests and some other small refactoring (playing better with MLIR).

**Comment by jnider** [comment]
For now, please keep going with what you have. Let's discuss the future of AIRBIN at the meeting on Tuesday and hopefully we'll come up with a good decision then.
-------- Original message --------From: Maksim Levental ***@***.***> Date: 2023-11-30  20:52  (GMT-08:00) To: Xilinx/mlir-aie ***@***.***> Cc: Joel Nider ***@***.***>, Mention ***@***.***> Subject: Re: [Xilinx/mlir-aie] Implement target backend for AIRBIN (PR #668) 
@jnider thanks for the explanation. I expected as much but wasn't sure. So are you comfortable with this approach until we get the as-of-yet-uknown standards body on board? If so, I'll keep going with my plans for the tests and some other small refactoring (playing better with MLIR).

**Comment by makslevental** [comment]
@jnider @stephenneuendorffer @jgmelber @SamuelBayliss 

This is good to go on my side. The only remaining question is about why the `'.sdma.bd (2)'` and `'.tdma.ctl (1)'` and `'.prgm.mem (1)'` section remain unintialized (as reflected by the `Content:` sections in the `obj2yaml` dump).

File: .github/workflows/buildAndTest.yml
Comment (makslevental) [reply .github/workflows/buildAndTest.yml:85]: Because, for example, it's not trivial to get libelf on a MacBook (where myself and others often do dev). Note I'm being explicit in the CI yamls but this option is `ON` by default if the user has libelf (and CMake can find it).

File: .github/workflows/lintAndFormat.yml
Comment (makslevental) [reply .github/workflows/lintAndFormat.yml:224]: These are my CI scripts that I'm iterating on. This will get resolved when they stabilize in main.

File: test/Targets/AIETargetAirbin/core_6_2.elf
Comment (makslevental) [reply test/Targets/AIETargetAirbin/core_6_2.elf:1]: I strongly strongly strongly advise leaving this file in order to have as close a unit test as possible.

Comment (jgmelber) [reply test/Targets/AIETargetAirbin/core_6_2.elf:1]: I agree this is worth checking in for the test

Comment (makslevental) [reply test/Targets/AIETargetAirbin/core_6_2.elf:1]: Ideally both the core elf files and airbin generator could check for compatible versions (but I'm not advocating for doing that now).

File: lib/Targets/CMakeLists.txt
Comment (makslevental) [inline lib/Targets/CMakeLists.txt:23]: d'oh

**Review by eddierichter-amd — APPROVED** [review]
LGTM

File: .github/workflows/buildAndTest.yml
Comment (jnider) [reply .github/workflows/buildAndTest.yml:85]: ok, got it

File: lib/Targets/AIETargetAirbin.cpp
Comment (jnider) [reply lib/Targets/AIETargetAirbin.cpp:31]: the team agreed - this is the way to go
we are not interested in defining our own official ELF format at this time, so we will just stick with this define
This will likely be moved to a header file in the near future, but for now this is fine

**Pull request merged**
